### PR TITLE
Do not abort if deal is not found

### LINKED ISSUE #529: Deal not found by market when terminating sector in miner cron
Reported from Lotus. Here's the execution traces of [miner](https://ipfs.io/ipfs/QmUkeq4p7TNh9A73MTXUBgMrb1S6y53smgtVHnXnBYEaRU/exec.html#bafy2bzacednhgynrysnxavsjl3speafkrjcii7dhzcwnlegr4glma2wxxzovo-bafy2bzacebehebh5suaib574cpglh42koal2rcs4ovgmpyr3mslqnb5duzvs2-bafy2bzacecahyrcyy67oces4pr4hl7zezc5qchcgvrk54xvimbn3ayegb6zmm) and [market](https://ipfs.io/ipfs/QmUkeq4p7TNh9A73MTXUBgMrb1S6y53smgtVHnXnBYEaRU/exec.html#bafy2bzacednhgynrysnxavsjl3speafkrjcii7dhzcwnlegr4glma2wxxzovo-bafy2bzacebehebh5suaib574cpglh42koal2rcs4ovgmpyr3mslqnb5duzvs2-bafy2bzacecahyrcyy67oces4pr4hl7zezc5qchcgvrk54xvimbn3ayegb6zmm-bafy2bzacebzybfajffcs3ytfvwpyajdmk46zxtmbd7ckahjisbithzjprizwy) actors.

Either the wrong deal ID was stored, or a deal was removed from the market actor before the sector ended. Maybe that latter case is expected somehow and should not cause an abort.

### PR DESCRIPTION
**Pull request by aarshkshah1992**
Closes #529 

@anorth Should the scope of this PR expand to write tests for `OnMinerSectorsTerminate` too so we can test changes like these or do you want to wait out and roll it into the issue we have to write tests for the Market Actor.

### CODE REVIEW

File: actors/builtin/market/market_actor.go [hunk 1]
```diff
@@ -381,6 +382,11 @@ func (a Actor) OnMinerSectorsTerminate(rt Runtime, params *OnMinerSectorsTermina
 
 		for _, dealID := range params.DealIDs {
 			deal, err := proposals.Get(dealID)
+			// deal could have terminated and hence deleted before the sector is terminated.
+			// we should simply continue instead of aborting execution here if a deal is not found.
+			if errors.Unwrap(err) == errDealNotFound {
```
Comment (anorth) [inline actors/builtin/market/market_actor.go:387]: Gah, not your fault but this is the breaking straw for me. Please refactor `proposals.Get` to return three values `(deal, found bool, err error)`, distinguishing these cases without introducing error introspection patterns.

File: actors/builtin/market/market_actor.go [hunk 2]
```diff
@@ -2,6 +2,7 @@ package market
 
 import (
 	"fmt"
+	"github.com/pkg/errors"
```
Comment (anorth) [inline actors/builtin/market/market_actor.go:5]: FYI we are moving away from `pkg/errors`, don't introduce more use of it. Use `xerrors`.

Comment (aarshkshah1992) [reply actors/builtin/market/market_actor.go]: We don't this anymore.

File: actors/builtin/market/market_actor.go (hunk 1 above)
Comment (aarshkshah1992) [reply actors/builtin/market/market_actor.go]: This is done.

**Comment by aarshkshah1992** [comment]
@anorth I've made the changes. Please take a look.

File: actors/builtin/market/market_actor.go [hunk 3]
```diff
@@ -346,10 +346,13 @@ func (a Actor) ActivateDeals(rt Runtime, params *ActivateDealsParams) *adt.Empty
 				rt.Abortf(exitcode.ErrIllegalArgument, "deal %d already included in another sector", dealID)
 			}
 
-			proposal, err := proposals.Get(dealID)
+			proposal, found, err := proposals.Get(dealID)
 			if err != nil {
 				rt.Abortf(exitcode.ErrIllegalState, "get deal %v", err)
 			}
```
Comment (anorth) [inline actors/builtin/market/market_actor.go:352]: Nit: use `RequireNoErr` here (and include the deal id).

File: actors/builtin/market/market_actor.go [hunk 4]
```diff
@@ -346,10 +346,13 @@ func (a Actor) ActivateDeals(rt Runtime, params *ActivateDealsParams) *adt.Empty
...
+			if !found {
+				rt.Abortf(exitcode.ErrIllegalState, "dealId %d not found", dealID)
```
Comment (anorth) [inline actors/builtin/market/market_actor.go:354]: `ErrNotFound`.

File: actors/builtin/market/market_actor.go [hunk 5]
```diff
@@ -633,10 +642,13 @@ func ValidateDealsForActivation(st *State, store adt.Store, dealIDs []abi.DealID
 	totalDealSpaceTime := big.Zero()
 	totalVerifiedSpaceTime := big.Zero()
 	for _, dealID := range dealIDs {
-		proposal, err := proposals.Get(dealID)
+		proposal, found, err := proposals.Get(dealID)
 		if err != nil {
 			return big.Int{}, big.Int{}, fmt.Errorf("failed to load deal %d: %w", dealID, err)
 		}
+		if !found {
+			return big.Int{}, big.Int{}, fmt.Errorf("dealId %d not found", dealID)
```
Comment (anorth) [inline actors/builtin/market/market_actor.go:650]: Aside: I'm aware we'd like this to be ErrNotFound too, but can't propagate that information yet.

File: actors/builtin/market/market_state.go [hunk 6]
```diff
@@ -361,10 +361,13 @@ func (st *State) mustGetDeal(rt Runtime, dealID abi.DealID) *DealProposal {
...
 	}
 
-	proposal, err := proposals.Get(dealID)
+	proposal, found, err := proposals.Get(dealID)
 	if err != nil {
 		rt.Abortf(exitcode.ErrIllegalState, "get proposal: %v", err)
 	}
+	if !found {
+		rt.Abortf(exitcode.ErrIllegalState, "dealId %d not found", dealID)
+	}
```
Comment (anorth) [inline actors/builtin/market/market_state.go:370]: Use `RequireNoErr` and `ErrNotFound` similar to above.

File: actors/builtin/market/market_actor.go (hunk 3 above)
Comment (aarshkshah1992) [reply actors/builtin/market/market_actor.go]: This is done.

File: actors/builtin/market/market_actor.go (hunk 4 above)
Comment (aarshkshah1992) [reply actors/builtin/market/market_actor.go]: This is done.

File: actors/builtin/market/market_state.go (hunk 6 above)
Comment (aarshkshah1992) [reply actors/builtin/market/market_state.go:404]: This is done.

**Pull request merged**